tests/test_gpu_examples.py

```diff
@@ -944,6 +944,51 @@ def test_causal_lm_training_multi_gpu(self):
 
 
 @require_torch_gpu
+class OffloadSaveTests(unittest.TestCase):
+    def setUp(self):
+        self.causal_lm_model_id = "gpt2"
+
+    def tearDown(self):
+        r"""
+        Efficient mechanism to free GPU memory after each test. Based on
+        https://github.com/huggingface/transformers/issues/21094
+        """
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @pytest.mark.single_gpu_tests
+    @require_torch_gpu
+    def test_offload_merge(self):
+        r"""
+        Test merging and unloading of a model with offloaded modules.
+        """
+        torch.manual_seed(0)
+        model = AutoModelForCausalLM.from_pretrained(self.causal_lm_model_id)
+        tokenizer = AutoTokenizer.from_pretrained(self.causal_lm_model_id)
+        memory_limits = {0: "0.4GIB", "cpu": "5GIB"}
+        # offloads around half of all transformer modules
+        device_map = infer_auto_device_map(model, max_memory=memory_limits)
+        self.assertTrue(0 in device_map.values())
+        self.assertTrue("cpu" in device_map.values())
+
+        config = LoraConfig(task_type="CAUSAL_LM", init_lora_weights=False, target_modules=["c_attn"])
+
+        model = get_peft_model(model, config)
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            model.save_pretrained(tmp_dir)
+            # load the model with device_map
+            model = AutoModelForCausalLM.from_pretrained(self.causal_lm_model_id, device_map=device_map).eval()
+            self.assertTrue(len({p.device for p in model.parameters()}) == 2)
+            model = PeftModel.from_pretrained(model, tmp_dir, max_memory=memory_limits)
+
+        input_tokens = tokenizer.encode("Four score and seven years ago", return_tensors="pt")
+        pre_merge_olayer = model(input_tokens)[0]
+        model = model.merge_and_unload()
+        post_merge_olayer = model(input_tokens)[0]
+
+        self.assertTrue(torch.allclose(pre_merge_olayer, post_merge_olayer))
```

Comment:
Unfortunately, this test fails for me:

```python
>>> torch.testing.assert_close(pre_merge_olayer, post_merge_olayer)
*** AssertionError: Tensor-likes are not close!

Mismatched elements: 301501 / 301542 (100.0%)
Greatest absolute difference: 2.2849273681640625 at index (0, 0, 11552) (up to 1e-05 allowed)
Greatest relative difference: 0.0675124004483223 at index (0, 0, 1862) (up to 1.3e-06 allowed)
```

Clearly, the values are highly correlated, but the difference is not negligible. Of course, we could set a higher tolerance to make the test pass, but I wonder if something fishy is going on here. When I run this completely on GPU, the assert passes.

When inspecting the differences, this is what I saw (sample of 0.1% of all values):

![image](https://github.com/huggingface/peft/assets/6229650/e8177298-9efc-4f4b-bb36-70cd4a329c16)

Do you have any idea what could be happening here?

Reply:
This is my fault, I pushed my test changes to the wrong branch (this was checking `model.merge_and_unload` which currently does not use `model.merge_adapter` and should not work). This has been corrected now!